tests/unit/charms/tls_certificates_interface/v4/test_tls_certificates_v4_requires.py

```diff
@@ -64,29 +78,110 @@ def get_certificate_secret(self, secrets: Iterable[Secret]) -> Secret:
 
     @pytest.fixture(autouse=True)
     def context(self):
-        self.ctx = scenario.Context(
+        self.ctx = testing.Context(
             charm_type=DummyTLSCertificatesRequirerCharm,
             meta=METADATA,
             config=METADATA["config"],
             actions=METADATA["actions"],
         )
 
-    def test_given_private_key_not_created_when_certificates_relation_created_then_private_key_is_generated(  # noqa: E501
+    def test_given_private_key_not_created_and_not_passed_when_certificates_relation_created_then_private_key_is_generated(  # noqa: E501
         self,
     ):
-        certificates_relation = scenario.Relation(
+        certificates_relation = testing.Relation(
             endpoint="certificates",
             interface="tls-certificates",
             remote_app_name="certificate-requirer",
         )
-        state_in = scenario.State(
+        state_in = testing.State(
             relations={certificates_relation},
             config={"common_name": "example.com"},
         )
 
         state_out = self.ctx.run(self.ctx.on.relation_created(certificates_relation), state_in)
 
         assert self.private_key_secret_exists(state_out.secrets)
+        secret = state_out.get_secret(label=f"{LIBID}-private-key-0")
+        assert secret.latest_content is not None
+        with open(
+            "tests/unit/charms/tls_certificates_interface/v4/dummy_requirer_charm/private_key.pem",
+            "r",
+        ) as f:
+            private_key = f.read()
+            assert private_key
+            assert private_key != secret.latest_content["private-key"]
+
+    @patch(
+        f"{BASE_CHARM_DIR}.get_private_key",
+        MagicMock(return_value=get_private_key_from_file()),
+    )
```

Comment:
I don't love the use of mocking when we can avoid it. It's sorta a last resort IMO.

I think ideally we'd have some way of interacting with the object (in this case, the charm) to change its behaviour. It's a little messy though, since we can't provide that in the constructor (ecause that is dictated by the framework, since these are in a way more integration test-like than unit tests). The best I can think of is an action or a config item.

Anyway, almost certainly not worth changing at this point, so feel free resolve, but I thought I'd highlight it anyway.

Reply:
Oops, I said _mocking_, I meant _patching_.